### ISSUE TITLE
Implement uncertainty

### ISSUE THREAD

**Issue by EvanMachusak** [FirelyTeam/firely-cql-sdk#132]
We don't implement uncertainty in the same way as the other two open source implementations, and this will cause testing differences that would be otherwise unresolvable.

We need to replace the use of `int?`, `long?`, and `decimal?` as the default bound types for the ELM `System.Integer`, `System.Long`, and `System.Decimal` types respectively to CQL types like `CqlInteger`, `CqlLong`, and `CqlDecimal`.  These new structures need to contain their primitive value, and must also contain a property that can contain the uncertainty range.  When the uncertainty range is needed and is populated, the result value is considered *uncertain*.  Uncertainty will need to be taken into account in a myriad of operators, mainly boolean comparison operators.

These types should be implicitly convertible `from` their respective primitive types.  It is not clear whether they should be implicitly convertible `to` their primitive types, since when they are uncertain, we would be collapsing a range down to a scalar value.  It may be reasonable to return `null` in such cases, or throw a runtime exception.  It may also be preferable to make the cast possible but explicit.

We will need comparers and type converters implemented for these new types.

The signatures of all of the `ICqlOperators` interface will need to have their primitive return types changed to use these structures so that uncertainty can be propagated.

This is a significant change and will result in a large pull request.

**Comment by ewoutkramer** [comment]
BTW: this will be added to the language as an official Uncertain<T> type, so I propose we do not act until that's official.

**Comment by alexzautke** [comment]
# Proposal: scoped uncertainty type (revised approach)

This proposal supersedes the original sketch above in one key respect: rather than replacing `int?`/`long?`/`decimal?` wholesale across `ICqlOperators`, it introduces a dedicated uncertainty type threaded **statically, only where uncertainty can actually flow**. The original proposal predates the typed-IR pipeline; now that every expression has a concrete static C# type end-to-end, scoped static propagation is both feasible and much cheaper than a wholesale primitive replacement.

## 1. Grounding

Per spec §5 (Language Semantics, "Uncertainty"): *"the concept of uncertainty is not exposed directly in CQL or in ELM, but is defined as an implementation detail."* An uncertainty is a closed interval with precisely specified comparison semantics — e.g. equality is defined as:

```
if A.low <= B.high and A.high >= B.low
  then if A.low = A.high and B.low = B.high then true else null
  else false
```

and relational comparison analogously (`A.high < B.low → true; A.low >= B.high → false; else null`), including the boundary case where overlapping-by-one still yields a decidable answer.

Uncertainty is **born in exactly one place**: `DurationBetween` / `DifferenceBetween` over `Date`/`DateTime`/`Time` operands whose precision is insufficient for the requested unit (`months between DateTime(2005) and DateTime(2006, 7)` → `[16, 44]`). Today these return `int?` and silently produce a point value; the engine answers `19` where the conformance suite expects `[18, 49]`.

Currently blocked on this (skipped in the DQIC XML suite): `DateTimeDurationBetweenUncertainAdd/Subtract/Multiply/Interval/Interval2`, `DateTimeDurationBetweenMonthUncertain2`, `DateTimeDurationBetweenYear`, `DateTimeUncertain` — plus `DateTimeDurationBetweenUncertainDiv`, which is marked `invalid="true"` (the suite expects division to *error*, a quirk to confirm during the spike).

## 2. Design

**New type** (Cql.Abstractions/Primitives): `CqlUncertainty<T>` — closed interval `(Low, High)` over an ordered point type, in practice `int`.
- Implicit conversion **from** point value (degenerate uncertainty, `low == high`).
- No implicit conversion **to** the point type. Explicit `Collapse()`/`TryGetPointValue()`: point value when degenerate, `null` otherwise. (This answers the original issue's open question: collapsing a genuine range to a scalar loses information, so it never happens implicitly.)
- Comparison semantics exactly per the spec pseudocode above, implemented as operators returning `bool?` plus a comparer that plugs into `CqlComparers`. Equivalence is never null: overlap-but-not-decidable maps to `false` (consistent with the null-comparison→false mapping `CqlComparisonToEquivalence` got in #1417/PR #1425).
- Interval arithmetic: `+`, `-`, `*` combine bounds (`[16,44] + [16,44] = [32,88]`, matching the conformance expectations); `div`/`mod` behavior confirmed in the spike (see quirk above).

**Birth sites** (Cql.Runtime): the six `DurationBetween`/`DifferenceBetween` overloads change return type from `int?` to `CqlUncertainty<int>?`. When precision suffices, the result is degenerate; callers that only ever see degenerate values behave exactly as today.

**Static propagation** (Cql.Compiler / CqlOperatorsBinder): the binder already resolves `ICqlOperators` overloads from static operand types. New overloads are added for the operator closure reachable from an uncertain operand:
- arithmetic: `Add/Subtract/Multiply(CqlUncertainty<int>?, CqlUncertainty<int>?)` (mixed point/uncertain unifies via the implicit conversion)
- comparisons: the six relational/equality operators
- membership: `In(CqlUncertainty<int>?, CqlInterval<int?>?)` and `between`
- `Abs`, `Negate` (cheap, likely needed)

Everything else — aggregates, list membership of uncertainties, `ToString` — is **explicitly deferred** until a conformance test or real measure demands it; the binder failing to resolve an unimplemented combination is a compile-time error, not a silent wrong answer, which makes the deferral safe.

**ELM/CQL surface: none.** The uncertainty type never appears in ELM or model info; it exists only in the runtime and the generated C#. This matches the spec's "implementation detail" framing and means the FHIR type converters are untouched (uncertainty cannot cross a retrieve boundary) — a significant scope reduction vs. the original proposal.

## 3. Delivery plan (4 stages, each independently shippable)

1. **Spike** (no product code): enumerate the closure the 10 skipped tests + spec §5 actually require; confirm the `div` error quirk; inventory which HEDIS/CMS measures contain duration expressions over potentially-imprecise dates (blast-radius estimate for stage 2). Output: a checklist that fixes stage 3's scope.
2. **PR 1 — foundations**: `CqlUncertainty<T>`, arithmetic, comparer, exhaustive unit tests derived from the spec pseudocode (incl. width-zero and boundary-overlap cases). No pipeline changes, no version bump.
3. **PR 2 — birth sites + propagation**: operator signature changes, binder overloads, generated-code changes. This is the breaking PR: **major `GeneratorToolVersion` bump**, new/updated `LibraryInstanceInvoker` support check, regenerate every checked-in `*.g.cs`, breaking-change release note. Point-precision behavior must be bit-identical (degenerate uncertainties collapse in comparisons/arithmetic to today's answers) — HEDIS Golden Parity and CMS integration suites are the gate, exactly as they caught the #1417 throw-vs-null mistake.
4. **PR 3 — conformance**: unskip the ~10 XML tests. Needs one harness decision: the suite's expectations compare a duration result against `Interval[16, 44]`, which doesn't type-check in strict CQL — the harness (like the Java engine) needs an internal equality between `CqlUncertainty<int>` and `CqlInterval<int?>` for assertion purposes only.

## 4. Open questions

1. **Propagation cut-off**: is compile-error-on-unimplemented-combination acceptable for v1 (my position: yes — loud and safe), or must every operator combination work before shipping?
2. **`Long`/`Decimal` uncertainties**: duration results are `Integer`; is `CqlUncertainty<int>` alone sufficient for v1? (I believe yes; the struct stays generic so widening later is additive.)
3. **`DateTimeDurationBetweenUncertainDiv`** expects an error — confirm against spec + Java engine behavior during the spike before deciding whether `div` propagates or signals.

Estimated shape: spike is small; PR 1 is small and riskless; PR 2 is the big one (comparable to a codegen-version release); PR 3 is small once 1–2 land.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by baseTwo** [comment]
## Still relevant — keeping open

Reviewed as part of a sweep over the oldest open issues (this one is from 2023). Rather than close it on age, here is what I checked and what is actually left.

### Assessment

Nothing has been implemented and the need is unchanged. There is no uncertainty type anywhere in the SDK (a repo-wide grep for "uncertain" finds only doc comments about null-on-imprecise-comparison, no `CqlUncertainty`/`Uncertain<T>` type). The birth sites named by the spec still return point values: `DurationBetween` and `DifferenceBetween` over Date/DateTime/Time are all `int?` on ICqlOperators. Nine DQIC conformance tests are still explicitly skipped with the reason "We don't support uncertainty". The issue is also actively being worked as a design: alexzautke posted a full revised proposal on 2026-07-26 (scoped `CqlUncertainty<T>` threaded statically through the typed-IR pipeline, 4-stage delivery plan, 3 open questions) which supersedes the body's original wholesale int?/long?/decimal? replacement. ewoutkramer's 2024 comment ("wait until Uncertain<T> is official in the language") is worth revisiting - spec 5 explicitly frames uncertainty as an implementation detail not exposed in CQL or ELM, which is the premise of the newer proposal. What remains: everything - the type, the comparison/arithmetic semantics, the operator signature changes at the birth sites, binder overloads (a breaking change requiring a major GeneratorToolVersion bump and regenerated *.g.cs), and unskipping the 9 conformance tests. Suggest the maintainer update the issue body to the revised design so the stale approach in the body does not mislead whoever picks it up.

### Evidence

Cql/CqlToElmTests/(tests)/SkippedTests.cs:42-50 - nine entries (DateTimeDurationBetweenMonthUncertain2, DateTimeDurationBetweenUncertainAdd/Div/Interval/Interval2/Multiply/Subtract, DateTimeDurationBetweenYear, DateTimeUncertain) all with reason "We don't support uncertainty". Cql/Cql.Runtime/Operators/ICqlOperators.cs:227-229 and :233-235 - DifferenceBetween/DurationBetween still return `int?`. No CqlUncertainty/Uncertain type exists (grep -ril "uncertain" over Cql/, CoreTests/, docs/ returns only XML-doc mentions in ICqlComparer.cs, CqlDate/CqlDateTime/CqlTime.cs and a ListOperators comment). GitHub search for open uncertainty PRs returns none. Spec basis: spec/cql/condensed/05-languagesemantics.md:579-594 (uncertainty is a closed interval, defined as an implementation detail).

### Does it still reproduce?

Yes for the observable symptom: duration between imprecise dates still silently yields a point value instead of an uncertainty range, which is why the nine conformance tests above remain skipped rather than passing. Confirmed by reading the current ICqlOperators signatures and the skip list; I did not need to run the suite since the skips are declarative.

---

_Verified against `develop` at the time of writing; confidence: **high**. If any of the above is wrong, the issue is a better candidate for closing than it looks._
